falcon/cyutil/streams.pyx

```diff
@@ -0,0 +1,312 @@
+from libc.stdint cimport uint32_t
+
+import functools
+import io
+
+DEFAULT_CHUNK_SIZE = io.DEFAULT_BUFFER_SIZE * 4
+
+
+cdef class BufferedStream:
+
+    cdef _read_func
+    cdef Py_ssize_t _chunk_size
+
+    cdef bytes _buffer
+    cdef Py_ssize_t _buffer_len
+    cdef Py_ssize_t _buffer_pos
+    cdef Py_ssize_t _max_bytes_remaining
+
+    def __cinit__(self, read, max_stream_len, chunk_size=None):
+        self._read_func = read
+        self._chunk_size = chunk_size or DEFAULT_CHUNK_SIZE
+
+        self._buffer = b''
+        self._buffer_len = 0
+        self._buffer_pos = 0
+        self._max_bytes_remaining = max_stream_len
+
+    def peek(self, Py_ssize_t amount=-1):
+        cdef Py_ssize_t read_amount
+
+        amount = int(amount)
+        if amount < 0 or amount > self._chunk_size:
+            amount = self._chunk_size
+
+        if self._buffer_len - self._buffer_pos < amount:
+            read_amount = (self._chunk_size - self._buffer_len +
+                           self._buffer_pos)
+            if read_amount >= self._max_bytes_remaining:
+                read_amount = self._max_bytes_remaining
+
+            if self._buffer_pos == 0:
+                self._buffer += self._read_func(read_amount)
+                self._buffer_len = len(self._buffer)
+            else:
+                self._buffer = (self._buffer[self._buffer_pos:] +
+                                self._read_func(read_amount))
+                self._buffer_len = len(self._buffer)
+                self._buffer_pos = 0
+            self._max_bytes_remaining -= read_amount
+
+        return self._buffer[self._buffer_pos:self._buffer_pos + amount]
+
+    cdef Py_ssize_t _normalize_size(self, amount):
+        cdef Py_ssize_t result
+        cdef Py_ssize_t max_amount = self._max_bytes_remaining + self._buffer_len
```

Comment:
This was not fixed in comparison to the Python version (missing support for `self._buffer_pos`).
I'll fix and add more tests.